tests/test_application.py

```diff
@@ -0,0 +1,1611 @@
+            else:
+                assert app.job_queue is None
+            # app.start() should not start the updater!
+            assert not app.updater.running
+            await asyncio.sleep(0.05)
+            assert app.update_queue.empty()
+            assert self.received == 1
+
+            await app.updater.start_polling()
+            await app.stop()
+            assert not app.running
+            # app.stop() should not stop the updater!
+            assert app.updater.running
+            if job_queue:
+                assert not app.job_queue.scheduler.running
+            else:
+                assert app.job_queue is None
+            await app.update_queue.put(2)
+            await asyncio.sleep(0.05)
+            assert not app.update_queue.empty()
+            assert self.received != 2
+            assert self.received == 1
+
+            await app.updater.stop()
+
+    @pytest.mark.asyncio
+    async def test_error_start_stop_twice(self, app):
+        async with app:
+            await app.start()
+            assert app.running
+            with pytest.raises(RuntimeError, match='already running'):
+                await app.start()
+
+            await app.stop()
+            assert not app.running
+            with pytest.raises(RuntimeError, match='not running'):
+                await app.stop()
+
+    @pytest.mark.asyncio
+    async def test_one_context_per_update(self, app):
+        self.received = None
+
+        async def one(update, context):
+            self.received = context
+
+        def two(update, context):
+            if update.message.text == 'test':
+                if context is not self.received:
+                    pytest.fail('Expected same context object, got different')
+            else:
+                if context is self.received:
+                    pytest.fail('First handler was wrongly called')
+
+        async with app:
+            app.add_handler(MessageHandler(filters.Regex('test'), one), group=1)
+            app.add_handler(MessageHandler(filters.ALL, two), group=2)
+            u = make_message_update(message='test')
+            await app.process_update(u)
+            self.received = None
+            u.message.text = 'something'
+            await app.process_update(u)
+
+    def test_add_handler_errors(self, app):
+        handler = 'not a handler'
+        with pytest.raises(TypeError, match='handler is not an instance of'):
+            app.add_handler(handler)
+
+        handler = MessageHandler(filters.PHOTO, self.callback_set_count(1))
+        with pytest.raises(TypeError, match='group is not int'):
+            app.add_handler(handler, 'one')
+
+    @pytest.mark.asyncio
+    @pytest.mark.parametrize('group_empty', (True, False))
+    async def test_add_remove_handler(self, app, group_empty):
+        handler = MessageHandler(filters.ALL, self.callback_increase_count)
+        app.add_handler(handler)
+        if not group_empty:
+            app.add_handler(handler)
```

Comment:
is this as intended?

Reply:
yes - I want to check that corresponding list is deleted from `app.handlers` only if the group is empty after removing the handler